### PR TITLE
fix(bootstrap): `deploy-role` could directly access buckets in target account

### LINKED ISSUE #12985: (cli): reduce deploy role scope
Currently the deploy role has a lot of access to *:

https://github.com/aws/aws-cdk/blob/b96efd8aaa143845b9fe315a9ee1e8398c4d83c2/packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml#L314-L336

I wonder if we can scope this down a bit by qualifier, so one bootstrap instance can't access things from another? Maybe with tags or enforcing a naming scheme? E.g. all stacks must start with `cdk-${qualifier}-`

### Use Case

We're investigating bootstrapping for each service with different qualifiers to try and reduce the blast radius if access to the deployment role (and others) from one bootstrap instance is compromised. The other roles/policies look like they're scoped pretty well. The `FilePublishingRoleDefaultPolicy` for example can only be used with the `StagingBucket.Arn` for that bootstrap instance.

I think we'll need to create a custom deploy role which can only invoke specific CloudFormation stacks with a restricted execution policy (unless this could be an official feature of the bootstrap stack in the future), but was looking to use this as a template, and if this level of access is really needed, I think the isolation we want really isn't going to be possible.

Thanks.

### Proposed Solution

I'm not sure if we can actually scope the permissions down from `*`. The comment

```
# Permissions needed in CodePipeline.
# S3 and KMS are needed on '*',
# as we don't know the IDs of the CodePipeline's bucket and key in advance
```

makes it sound like we can't, but I'm really hoping that's not the case.

* [x] :wave: I may be able to implement this feature request (depending what the solution ends up being)
* [x] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by rix0rrr**
In the previous formulation of the permissions of the `deploy-role`, it
had `s3:GetObject` on `*`, meaning it could read buckets in the target
account.

The permissions were only necessary to read cross-account artifact
buckets created by CDK Pipelines, and so strictly speaking too much
data could be read.

While not strictly an escalation of privileges (given that the user
who is allowed to assume `deploy-role` can already do *anything* to
the account using CloudFormation, and so is in principle already
able to read those buckets via a detour) it is still good practice
to lock the unnecessary direct permissions down.

The new permissions for the `deploy-role` as as follows:

* Can read its own staging bucket (necessary to read the initial
  CloudFormation template for `cdk deploy`)
* Can read S3 buckets *not* in its own account: this enables
  cross-account deployments (by reading the artifact bucket in the
  other account). The nature of cross-account access
  *requires* that the target bucket has a trust policy allowing the
  access as well, so not an exploit.
* Can use all KMS keys as long as it's via S3: makes sure that it
  cannot use KMS keys directly.
* Added `CreateStack` and `UpdateStack` for potential future deployment
  methods that don't use Change Sets.

In-account bucket accesses are not directly permitted by the principal
roles; however, in-account accesses are enabled by resource policy
statements on the buckets themselves (which are sufficient to extend
the required permissions to the principal).

Fixes #12985, closes #14082, closes #13422.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml [hunk 1]
```diff
@@ -383,30 +383,48 @@ Resources:
...
         - PolicyDocument:
             Statement:
-              - Action:
-                  # Permissions needed in CodePipeline.
-                  # S3 and KMS are needed on '*',
-                  # as we don't know the IDs of the CodePipeline's bucket and key in advance
+              - Sid: CloudFormationPermissions
+                Effect: Allow
+                Action:
                   - cloudformation:CreateChangeSet
```
Comment (eladb) [inline packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:390]: Can you please add CreateStack and UpdateStack?

**Review by eladb — CHANGES_REQUESTED** [review]

**Review by skinny85 — APPROVED** [review]
I really hope you tested these changes well 😉

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml (hunk 1 above)
Comment (NetaNir) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:390]: @eladb the `CreateChangeSet` allows creating an updating stacks by specifying `ChangeSetType ` as `Update`. See docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CreateChangeSet.html

This should be sufficient, unless Im missing some use case

Comment (eladb) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:390]: In some scenarios (eg the GH action I've been using to implement the CDK pipelines backend for GitHub) we may _need_ to use CreateStack and UpdateStack

**Review by eladb — APPROVED** [review]

**Comment by mrpackethead** [comment]
This seems sensible, and I cant' see anything that will cause a problem.

File: packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml [hunk 2]
```diff
@@ -385,30 +385,50 @@ Resources:
...
                   - s3:List*
                   - s3:Abort*
                   - s3:DeleteObject*
                   - s3:PutObject*
-                  # Necessary to write to the cross-region artifact replication bucket
-                  # https://aws.amazon.com/premiumsupport/knowledge-center/codepipeline-deploy-cloudformation/.
+                Resource: "*"
+                Condition:
+                    StringNotEquals:
+                        s3:ResourceAccount:
```
[the remark this answers is not in the document]
Comment (bkmnaidu) [reply packages/aws-cdk/lib/api/bootstrap/bootstrap-template.yaml:416]: I am running into the same issue as described by you @jamesorlakin as in the changes do not work for same-account deployments. Is there any update on this?

**Pull request merged**